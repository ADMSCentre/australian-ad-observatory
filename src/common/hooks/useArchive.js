import { useCallback, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import actions from 'common/state/archive/actions';

const useArchive = () => {
  const { items, index, loading, saving, ...archive } = useSelector(({ archive }) => archive);

  const oldestDate = index.reduce((acc, elem) => {
    const datetime = items[elem].observedAt;
    return acc <= datetime ? acc : datetime;
  }, Infinity);

  const oldest = oldestDate === Infinity ? undefined : oldestDate;

  const dispatch = useDispatch();

  const loadArchive = useCallback(
    /* istanbul ignore next */
    () => dispatch(actions.loadArchive()),
    [dispatch]
  );

  useEffect(() => {
    /* istanbul ignore else */
    if (!index.length && !loading && !saving) loadArchive();
  }, [index.length]);

  return { items, index, loading, saving, oldest, ...archive };
};

export default useArchive;
