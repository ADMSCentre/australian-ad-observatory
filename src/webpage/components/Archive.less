.ati-item {
  font-size: 11px;
  direction: ltr;
  display: block;

  &:hover {
    // background-color: rgba(29, 33, 41, 0.04);
  }
}

.post-wrapper {
  max-width: 1024px;
}
