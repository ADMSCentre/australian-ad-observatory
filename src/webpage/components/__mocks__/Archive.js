/* eslint-disable react/prop-types */
import React from 'react';

const Archive = () => <div id="archive-list">Archive stuff here</div>;

export default Archive;
