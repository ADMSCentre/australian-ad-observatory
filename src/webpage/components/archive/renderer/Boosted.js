import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faRocket } from '@fortawesome/free-solid-svg-icons';

const Boosted = ({ boosted }) =>
  boosted ? (
    <div className="btn btn-outline-success">
      <FontAwesomeIcon icon={faRocket} />
      Post was boosted
    </div>
  ) : null;

Boosted.displayName = 'Boosted';

Boosted.propTypes = { boosted: PropTypes.bool };

Boosted.defaultProps = { boosted: false };

export default Boosted;
