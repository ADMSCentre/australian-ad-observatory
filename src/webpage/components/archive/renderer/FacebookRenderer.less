.fb-item {
  padding: 0;
  width: 500px;
  background-color: #fff;
  border-radius: 5px;
  border: 1px solid #ddd;
  margin-top: 25px;
  margin-bottom: 15px;
  margin-right: 15px;

  &:first-child {
    padding: 12px 12px 0 12px;
  }

  .hidden_elem {
    display: none;
  }

  [data-testid="post_message"] {
    padding: 10px 0;
  }

  & > img {
    max-width: 90%;
    border: 0;
    margin: 0;
    padding: 0;
  }

  video {
    display: none;
  }

  div.mtm div div div span div img {
    width: 498px;
    height: auto;
    margin-left: -12px;
    border-top: 1px solid #ccc;
    border-bottom: 1px solid #ccc;
  }

  // Hide divs that have been otherwise rendering as stray periods.
  ._6spk {
    display: none;
  }

  // Don't allow user to click buttons or links.
  // a,
  // button {
  //   // pointer-events: none;
  // }

  h5 {
    font-size: 14px; // Match size of source bar w/ Facebook's size
    margin-bottom: 0; // Bring timestamp closer to the source bar title
  }

  // Add "arrow" in place of "to" in source bar.
  ._gb8.img::after {
    content: ">";
    font-style: normal;
  }

  // Remove word "to" in source bar.
  ._gb8 u {
    display: none;
  }

  ._6m6 {
    padding-top: 2px;
    margin-top: 3px;
    display: block;
    float: left;
    padding-bottom: 10px;

    a {
      color: #1d2129;
      font-size: 16px;
      font-weight: 600;

      &:hover {
        text-decoration: none;
        color: #1d2129;
      }
    }
  }

  // Bring author names inline with profile pics.
  ._6a._6b {
    display: inline-block;
    height: 20px !important;

    a {
      color: #385898;
      font-size: 14px;
      font-weight: 600;
      font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
      -webkit-font-smoothing: subpixel-antialiased;
    }
  }

  ._ohe ._6a {
    display: block;
    float: right;
    padding-left: 18px;
    height: 20px;
    vertical-align: middle;

    a {
      border: 1px solid rgb(190, 195, 201);
      background-color: rgb(238, 240, 242);
      color: #4b4f56;
      padding: 5px 8px;
      border-radius: 2px;

      &:hover {
        text-decoration: none;
      }
    }
  }

  ._6m7 {
    margin-top: 3px;
    color: #606770;
    font-size: 14px;
    line-height: 20px;
    word-break: break-word;
    display: none;
    overflow-x: hidden;
  }

  div[data-testid="post_message"] p {
    color: #1d2129;
    font-size: 14px;
    font-weight: 400;
    margin-bottom: 5px;
  }

  .accessible_elem {
    display: none;
  }

  ._4r_y {
    height: 2px;
  }

  // Hide inconsequential artifacts.
  ._52c6 {
    display: none;
  }

  div ._3x-2 {
    margin: 0;
    padding: 0;
  }

  // Hide buttons.
  button {
    display: none;
  }

  // Profile pic wrapper.
  ._5pb8._1yz2._8o._8s.lfloat._ohe,
  ._38vo {
    display: block;
    float: left;
    width: 32px;
    margin-right: 8px;
    border-radius: 16px;
    overflow: hidden;
  }

  ._38vo {
    max-height: 32px;
  }

  ._38vo img {
    max-height: 32px;
  }

  ._5x46,
  ._3dp ._29k {
    /* Profile name wrapper */
    margin-bottom: 6px;
  }

  ._3ekx {
    display: block;
  }

  ._6lz {
    padding-top: 10px;
    color: #606770;
    font-size: 12px;
  }

  .uiScaledImageContainer {
    max-width: 100%;
    width: auto;
    height: auto;
  }

  .scaledImageFitWidth {
    // max-width: 100%;
    // width: auto;
    // height: auto;
  }

  .ellipsis {
    line-height: 14px;
    text-transform: uppercase;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .card-title {
    display: none;
  }

  .card-subtitle {
    display: none;
  }

  &.fb-publicUserPost {
    // background-color: #e1edec;
    div.mtm div div div span div div {
      display: none;
    }
  }

  &.fb-publicPagePost {
    // background-color: #e1edec;
    div.mtm div div div span div div {
      display: none;
    }
  }

  &.fb-sponsoredPost {
    // background-color: #f1edec;

    a[role="button"] {
      //border: 1px solid;
      border-radius: 2px;
      box-sizing: content-box;
      font-size: 12px;
      -webkit-font-smoothing: antialiased;
      font-weight: bold;
      justify-content: center;
      padding: 0 8px;
      position: relative;
      text-align: center;
      text-shadow: none;
      vertical-align: middle;
    }
  }

  ul.uiList {
    margin-left: -52px;
  }

  ul.uiList li {
    display: none;
  }

  ul.uiList li:first-child {
    display: inline-block;
    border-top: 1px solid #ccc;
    width: 100%;
  }

  ul.uiList li img {
    width: 498px;
    height: auto;
  }

  ._5eut,
  ._42gt {
    display: none;
  }

  ._4lpe,
  ._3htz,
  ._5mly,
  ._40ja,
  ._45oh {
    display: none;
  }
}

.targeting {
  margin-left: 5px;
  width: 500px;

  h4 {
    font-size: 12px;
    font-weight: 600;
    margin: 0;
    padding: 0;
  }

  .reason {
    margin-right: 10px;

    .reason-name {
      padding-left: 5px;
    }
  }
}
