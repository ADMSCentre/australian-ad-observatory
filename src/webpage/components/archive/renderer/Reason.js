import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faUsers,
  faGlobe,
  faTransgenderAlt,
  faMapMarker,
  faDotCircle
} from '@fortawesome/free-solid-svg-icons';

const REASONS = {
  WAISTUICustomAudienceType: {
    name: 'Custom Audience',
    icon: faUsers
  },
  WAISTUILocaleType: {
    name: 'Locale',
    icon: faGlobe
  },
  WAISTUIAgeGenderType: {
    name: 'Age and Gender',
    icon: faTransgenderAlt
  },
  WAISTUILocationType: {
    name: 'Location',
    icon: faMapMarker
  }
};

const Reason = ({ reason }) =>
  REASONS[reason] ? (
    <div className="reason badge badge-primary">
      <FontAwesomeIcon icon={REASONS[reason].icon} />
      <span className="reason-name">{REASONS[reason].name}</span>
    </div>
  ) : (
    <div className="reason badge badge-primary">
      <FontAwesomeIcon icon={faDotCircle} />
      <span className="reason-name">{reason}</span>
    </div>
  );

Reason.displayName = 'Reason';

Reason.propTypes = {
  reason: PropTypes.string.isRequired
};

export default Reason;
