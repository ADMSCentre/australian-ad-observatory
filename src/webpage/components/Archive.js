import React, { Fragment, useState } from 'react';

import ArchiveHeader from './archive/ArchiveHeader';
import ArchiveTabs from './archive/ArchiveTabs';
import ArchiveList from './archive/ArchiveList';
import ArchiveLoading from './archive/ArchiveLoading';
import ArchiveEmpty from './archive/ArchiveEmpty';
import Pagination from './archive/Pagination';

import useArchive from 'common/hooks/useArchive';
import { observedAtRecentFirst } from '../utils/sorts';
import classifyItems from '../utils/classifyItems';

import './Archive.less';

const buildPerms = JSON.parse(process.env.PERMISSIONS);

const ALL_TABS = {
  facebookAds: {
    label: 'Facebook Ads',
    permission: 'FB_SHARE_SPONSORED_POSTS'
  },
  facebookUserPosts: {
    label: 'Facebook User Posts',
    permission: 'FB_SHARE_PUBLIC_USER_POSTS'
  },
  facebookPagePosts: {
    label: 'Facebook Posts',
    permission: 'FB_SHARE_PUBLIC_PAGE_POSTS'
  },
  youtubeAds: {
    label: 'YouTube Ads',
    permission: 'YT_SHARE_ADS'
  },
  youtubeRecommendations: {
    label: 'YouTube Recommendations',
    permission: 'YT_SHARE_RECOMMENDED_VIDEOS'
  },
  youtubeVideos: {
    label: 'YouTube Videos',
    permission: 'YT_SHARE_WATCHED_VIDEOS'
  }
};
const TABS_IN_USE = Object.keys(ALL_TABS).reduce((acc, elem) => {
  const tab = ALL_TABS[elem];
  if (buildPerms[tab.permission]) acc[elem] = tab.label;
  return acc;
}, {});

const ITEMS_PER_PAGE = 10;

const paginate = (items, pageNumber) =>
  items.slice((pageNumber - 1) * ITEMS_PER_PAGE, pageNumber * ITEMS_PER_PAGE);

const Archive = () => {
  const firstAllowedTab = Object.keys(TABS_IN_USE)[0];

  const { items, index, oldest, loading } = useArchive();
  const [activeTab, setActiveTab] = useState(firstAllowedTab);
  const [currentPage, setCurrentPage] = useState(1);

  if (!index.length) return loading ? <ArchiveLoading /> : <ArchiveEmpty />;
  const classified = classifyItems(items);

  const tabs = Object.keys(TABS_IN_USE).reduce((acc, elem) => {
    acc.push({
      name: elem,
      label: TABS_IN_USE[elem],
      active: elem === activeTab,
      badge: classified[elem].length
    });
    return acc;
  }, []);

  const sortedItems = [...classified[activeTab]].sort(observedAtRecentFirst);
  const pages = Math.ceil(sortedItems.length / ITEMS_PER_PAGE) /* istanbul ignore next */ || 1;
  const pagedItems = paginate(sortedItems, currentPage);

  /* istanbul ignore next */
  const changeTab = tab => {
    setActiveTab(tab);
    setCurrentPage(1);
  };

  return (
    <Fragment>
      <ArchiveHeader count={index.length} since={new Date(oldest)} />
      <ArchiveTabs tabs={tabs} onChange={changeTab} />
      <ArchiveList items={pagedItems} />
      <Pagination pages={pages} currentPage={currentPage} onChange={setCurrentPage} />
    </Fragment>
  );
};
Archive.displayName = 'Archive';

export default Archive;
