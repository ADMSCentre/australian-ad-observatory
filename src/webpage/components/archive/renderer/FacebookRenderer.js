import React from 'react';
import PropTypes from 'prop-types';

import Reason from './Reason';
import Boosted from './Boosted';

import './FacebookRenderer.less';

const itemShape = {
  id: PropTypes.string.isRequired,
  itemId: PropTypes.string.isRequired,
  itemType: PropTypes.string.isRequired,
  payload: PropTypes.shape({
    contentHtml: PropTypes.string.isRequired
  }).isRequired
};

const AdTargeting = ({ item }) => {
  if (!item.payload.adTargetingData) return null;

  const {
    data: { waist_targeting_data: targetingData, waist_is_marketplace_boosted_listing: boosted }
  } = item.payload.adTargetingData;
  const targetingReasons = targetingData.map(({ __typename }) => __typename);

  return (
    <div className="targeting">
      <h4>Ad Targeting</h4>
      <Boosted boosted={boosted} />
      {targetingReasons.map(reason => (
        <Reason key={reason} reason={reason} />
      ))}
    </div>
  );
};
AdTargeting.displayName = 'Ad targeting';

AdTargeting.propTypes = {
  item: PropTypes.shape(itemShape).isRequired
};

const FacebookRenderer = ({ item }) => {
  const {
    id,
    itemId: _itemId,
    itemType,
    payload: { contentHtml: html }
  } = item;

  const classes = `ati-item fb-item fb-${itemType}`;

  return (
    <div className="archive-item">
      <div id={id} className={classes} dangerouslySetInnerHTML={{ __html: html }} />
      <AdTargeting item={item} />
    </div>
  );
};

FacebookRenderer.propTypes = {
  item: PropTypes.shape(itemShape).isRequired
};

export default FacebookRenderer;
