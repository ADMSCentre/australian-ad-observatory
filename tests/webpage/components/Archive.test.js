import React, { useState as useStateMock } from 'react';

import renderer from 'react-test-renderer';

import useArchive from 'common/hooks/useArchive';

import Archive from 'webpage/components/Archive';

const observedAt = new Date('2020-02-02T00:00:00Z').getTime();

jest.mock('common/hooks/useArchive');
jest.mock('webpage/components/archive/ArchiveList');

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn()
}));

const id = '1';
const items = {
  [id]: {
    id,
    platform: 'youtube',
    itemType: 'recommendedVideo',
    hostVideo: { id: '12345', url: 'http://www.youtube.com' },
    observedAt
  },
  '2': {
    id: '2',
    platform: 'facebook',
    itemType: 'publicPagePost',
    observedAt
  },
  '3': { id: '3', platform: 'facebook', itemType: 'sponsoredPost', observedAt },
  '4': {
    id: '4',
    platform: 'facebook',
    itemType: 'publicUserPost',
    observedAt
  },
  '5': {
    id: '5',
    platform: 'facebook',
    itemType: 'publicPagePost',
    observedAt
  }
};

const index = [id, '2', '3', '4', '5'];
const setActiveTab = jest.fn();
const setCurrentPage = jest.fn();

let tree;

describe('facebook ads tab', () => {
  describe('when there is stuff in the archive', () => {
    beforeAll(() => {
      useStateMock
        .mockImplementationOnce(t => [t, setActiveTab])
        .mockImplementationOnce(p => [p, setCurrentPage]);
      useArchive.mockReturnValue({
        items,
        index,
        oldest: observedAt,
        loading: false
      });
      tree = renderer.create(<Archive />);
    });

    it('renders correctly', () => {
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when there is no stuff in the archive', () => {
    describe('when it is loading', () => {
      beforeAll(() => {
        useStateMock
          .mockImplementationOnce(t => [t, setActiveTab])
          .mockImplementationOnce(p => [p, setCurrentPage]);
        useArchive.mockReturnValue({ items: {}, index: [], loading: true });
        tree = renderer.create(<Archive />);
      });

      it('renders correctly', () => {
        expect(tree).toMatchSnapshot();
      });
    });

    describe('when it is simply empty', () => {
      beforeAll(() => {
        useStateMock
          .mockImplementationOnce(t => [t, setActiveTab])
          .mockImplementationOnce(p => [p, setCurrentPage]);
        useArchive.mockReturnValue({ items: {}, index: [], loading: false });
        tree = renderer.create(<Archive />);
      });

      it('renders correctly', () => {
        expect(tree).toMatchSnapshot();
      });
    });
  });
});

describe('youtube video tab', () => {
  describe('when there is stuff in the archive', () => {
    beforeAll(() => {
      useStateMock
        .mockImplementationOnce(() => ['youtubeVideos', setActiveTab])
        .mockImplementationOnce(p => [p, setCurrentPage]);
      useArchive.mockReturnValue({
        items,
        index,
        oldest: observedAt
      });
      tree = renderer.create(<Archive />);
    });

    it('renders correctly', () => {
      expect(tree).toMatchSnapshot();
    });
  });
});
