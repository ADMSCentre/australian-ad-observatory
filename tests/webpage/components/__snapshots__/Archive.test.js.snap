// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facebook ads tab when there is no stuff in the archive when it is loading renders correctly 1`] = `
<h3>
  Loading archive.
</h3>
`;

exports[`facebook ads tab when there is no stuff in the archive when it is simply empty renders correctly 1`] = `
<h3>
  You have nothing in your archive.
</h3>
`;

exports[`facebook ads tab when there is stuff in the archive renders correctly 1`] = `
Array [
  <div
    className="mb-2"
  >
    <div>
      This is an archive of the data this extension has collected from your social feeds.
    </div>
    <div
      className="small"
    >
      You have 
      5 items
       in your archive, collected since 
      <span
        className="as-date"
      >
        02/02/2020
      </span>
    </div>
  </div>,
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn border align-items-center btn-light"
      onClick={[Function]}
      type="button"
    >
      Facebook Ads
      <span
        className="badge badge-primary ml-2"
      >
        1
      </span>
    </button>
    <button
      className="btn border align-items-center btn-secondary"
      onClick={[Function]}
      type="button"
    >
      YouTube Videos
      <span
        className="badge badge-primary ml-2"
      >
        1
      </span>
    </button>
  </div>,
  <div
    id="archive-list"
  >
    Archive List stuff here
  </div>,
]
`;

exports[`youtube video tab when there is stuff in the archive renders correctly 1`] = `
Array [
  <div
    className="mb-2"
  >
    <div>
      This is an archive of the data this extension has collected from your social feeds.
    </div>
    <div
      className="small"
    >
      You have 
      5 items
       in your archive, collected since 
      <span
        className="as-date"
      >
        02/02/2020
      </span>
    </div>
  </div>,
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn border align-items-center btn-secondary"
      onClick={[Function]}
      type="button"
    >
      Facebook Ads
      <span
        className="badge badge-primary ml-2"
      >
        1
      </span>
    </button>
    <button
      className="btn border align-items-center btn-light"
      onClick={[Function]}
      type="button"
    >
      YouTube Videos
      <span
        className="badge badge-primary ml-2"
      >
        1
      </span>
    </button>
  </div>,
  <div
    id="archive-list"
  >
    Archive List stuff here
  </div>,
]
`;
